### DISCUSSION TITLE
Исходящие и входящие платежи в одной таблице из разных источников.

### DISCUSSION THREAD

**Discussion by EKhorikov** [totumonline/totum-ru-issues-and-discussions#114] [Как это сделать?] [answered]
Друзья, всем привет.
Нужна помощь!
Есть простая таблица "Платежи", в ней хранятся все ВХОДЯЩИЕ и ИСХОДЯЩИЕ (это ключевой момент вопроса) платежи.
![image](https://user-images.githubusercontent.com/92679628/164488757-547d86c0-e627-4836-8e7d-23535f4f73ba.png)
 Отличаются записи в этой таблице полем "Тип документа" и заполнением соответствующего поля "Приход" либо "Расход".
Вроде всё просто, но - входящие и исходящие платежи создаются в разных местах, чаще всего Исходящий из Заказа поставщику, а входящий из Проекта клиента. Или в "Проекте" есть "Входящий платёж" от клиента, но так же исходящий в виде расходов по этому "Проекту". Работает это через вкладки в цикле настроенные через linkToTable (открывается таблица через префильтр)
![image](https://user-images.githubusercontent.com/92679628/164487981-15dd94a7-e60a-4d13-91e2-df4fa5157e62.png)
Так вот задача:
Нужно учитывать оба типа платежей в одной таблице, но при формировании платежа из вкладки "Платежи по проекту" нужно чтоб по умолчанию было заполнено значение "Входящий платёж", при этом для заполнения, было доступно только поле "Приход", а "Расход" закрыто для редактирования. А при заполнении из вкладки "Расходы", наоборот - значение "Исходящий платёж" и доступно для редактирования поле "Расход".
Всё это нужно, для удобной аналитики Приход-Расход в одной таблице.
Заранее спасибо за помощь!

**Comment by erpic-ru** [comment]
Возможно, я не до конца понял задачу, но могу предположить такой вариант решения:)

На вкладках "Платежи по проекту" и "Расходы" не делать linktotable, а сделать расчетные в цикле. 

В хедере каждой вкладки разместить форму с кнопкой добавления транзакции в таблицу платежи (сумма, дата и др.) с нужными предустановленными параметрами (Тип платежа, номер проекта и другое) и пересчитывать таблицу после добавления

В табличной части делать выборку транзакций по проекту из "Платежей" и автозаполнять через insert.

**Comment by ttm-alexeypolunin** [comment, marked as answer]
Такие штуки делаются через скрытое от пользователей поле префильтра. Те когда вы используете link (вкладку) таблицы циклов пропишите в него передачу чекбокса например в скрытый параметр.

Дальше вы обращаетесь к этому префильтру в:

— строке добавления чтоб подставить значение
— форматировании чтоб заблокировать изменение
— коде действия чтоб заблокировать железобетонно

Здесь я при помощи этого параметра скрываю поля префильтра при открытии таблицы как вкладки цикла:

<img width="1493" alt="Снимок экрана 2022-04-21 в 19 28 53" src="https://user-images.githubusercontent.com/55755565/164508134-869faf8f-d990-4fcc-8685-e19cf83e2e5d.png">

**Comment by EKhorikov** [comment]
Для понимания, нарисовал небольшую схемку.
1. Есть два источника добавления документов в одну таблицу. Источник 1 вкладка с префильтром "ПРИХОД" источник 2 вкладка с префильтром "РАСХОД" .
2. Принцип добавления в обоих источников одинаковый, и в принципе это один и тот же документ, но необходимо: 
2.1 При добавлении из источника 1 "ПРИХОД блокировать для заполнения поле "Расход" и заполнять по умолчанию поле "Тип документа" значением "Входящий платёж"
2.2 При добавлении из источника 2 "РАСХОД" блокировать для заполнения поле "Приход" и заполнять по умолчанию поле "Тип документа" значением "Исходящий платёж"

![image](https://user-images.githubusercontent.com/92679628/164663020-0696d2db-5ac9-4736-8aff-69cca3d07f6d.png)

**Reply by ttm-alexeypolunin** [reply]
1. Добавляется скрытое поле в префильтр (видно только Создателю, но Редактируемое), например `fl_type`. Пусть `type1` = `Приход`, `type2` = `Расход`.
2. Передаем параметр в кнопке `link` для создания вкладки в цикле (для одной `type1` для другой `type2`):
```
=: linkToTable(...; filter: 'fl_type' = type1; ...)
``` 
3. В Тип документа, Приход и Расход пишем для автозаполнения нужного значения в Код:

```
=: if(condition: #fl_type != ""; then: $$type; else: "")

type: #fl_type

type1: "Исходящий платеж"
type2: "Входящий платеж"

```
Я здесь ветвление сделал через двойной доллар, вы можете попроще реализовать

4. В Форматировании на том же принципе ставим блокировку через `setFormat`

**Reply by EKhorikov** [reply]
@ttm-alexeypolunin, @erpic-ru,  спасибо за ответы! Буду пробовать.
